core/blockchain.go

```diff
@@ -281,9 +282,34 @@ func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, chainConfig *par
 	// Make sure the state associated with the block is available
 	head := bc.CurrentBlock()
 	if _, err := state.New(head.Root(), bc.stateCache, bc.snaps); err != nil {
-		log.Warn("Head state missing, repairing", "number", head.Number(), "hash", head.Hash())
-		if err := bc.SetHead(head.NumberU64()); err != nil {
-			return nil, err
+		// Head state is missing, before the state recovery, find out the
+		// disk layer point of snapshot(if it's enabled). Make sure the
+		// rewound point is lower than disk layer.
+		var diskRoot common.Hash
+		if bc.cacheConfig.SnapshotLimit > 0 {
+			diskRoot = rawdb.ReadSnapshotRoot(bc.db)
+		}
+		if diskRoot != (common.Hash{}) {
+			log.Warn("Head state missing, repairing", "number", head.Number(), "hash", head.Hash(), "snaproot", diskRoot)
+
+			var snapDisk uint64
+			if err := bc.SetHeadWithCondition(head.NumberU64(), func(block *types.Block, canStop *bool) {
```

Comment:
If I understand correctly, this will rewind up to the snapshot disk layer. But there's no guarantee we have a trie associated with that state. If we do not, we can't really stop there. Reads sure, can be served by the disk layer, but we can't write into a non-existet trie. Shouldn't we just go back to an existing trie, independent of the disk layer? I.e. it might make sense to capture the block number of the snapshot layer, but I don't think it's safe to stop setHead eraly. Am I missing something?

Reply:
Yes, we will rewind *at least* up to the snapshot disk layer.

> But there's no guarantee we have a trie associated with that state.

Yes it's also true. No guarantee the disk layer has an available trie. But if we ever meet the disk layer, the rewinding will continue until we find the available trie.
In this case, we will eventually find a trie(the worst case is genesis) and the disk layer is always higher than this block. So before the chain is catching up the disk layer, the entire snapshot is not available for using(read, write).

Reply:
Ah yes, the github diff collapsed 3 lines which contained a `continue`.